src/plugins/intel_cpu/src/nodes/memory.cpp

```diff
@@ -647,13 +758,31 @@ void MemoryInput::runDynamic(dnnl::stream strm) {
         memBlock->reset();
     }
 
-    //reshape output
-    const auto& newDims = processInitGraph ? getSrcMemoryAtPort(0)->getStaticDims() : stateDims;
+    MemoryPtr src = assignedMem;  // declare src memory
+    if (processInitGraph) {
+        if (haveSubgraph()) {
+            subGraph.ResetInferCount();
+            subGraph.Infer();
+            // depending on the memory sharing solution, we can return here if the memory is substituted from the
+            // external graph or override the src pointer with the memory pointer pointing to the subgraph output
+            // memory
+            auto& outputs = subGraph.GetOutputNodesMap();
+            OPENVINO_ASSERT(outputs.size() == 1);
+            auto itr = outputs.begin();
+            src = itr->second->getSrcMemoryAtPort(0);
```

Comment:
Because I had implemented shape inference, maybe I can remove them directly.

Reply:
They won't have different shapes if you apply changes suggested in https://github.com/openvinotoolkit/openvino/pull/26130#discussion_r1849391289, so that the memory object is shared between the node and the subgraph.

Reply:
I don't think so, I have updated codes based on https://github.com/openvinotoolkit/openvino/pull/26130#discussion_r1849391289
I got test result:
```
src shape: [1,1500,384], data ptr = 0x7f4ef41e84c0
assignedMem shape: [0,0,384], data ptr = 0x7f4ef41e84c0          (state)
```

Just add test code in the next https://github.com/xipingyan/openvino/blob/2bb7f7b62fc794e0e0cccbc2ebd14d761a1f5255/src/plugins/intel_cpu/src/nodes/memory.cpp#L774:
```
 auto& childEdges = getChildEdges();
            for (size_t i = 0; i < getOriginalOutputsNumber(); i++) {
                const auto mem = getDstMemoryAtPort(i);
                for (size_t j = getOriginalOutputsNumber(); j < childEdges.size(); j++) {
                    auto& childEdge = childEdges[j];
                    auto childEdgePtr = childEdge.lock();
                    assert(childEdgePtr);

                    if (childEdgePtr->getInputNum() == static_cast<int>(i)) {
                        childEdgePtr->getMemoryPtr()->redefineDesc(mem->getDescPtr());
                    }
                }
            }
            auto vec2str = [](ov::intel_cpu::VectorDims vec) {
                std::string str = "[";
                for (auto itm : vec) {
                    str = str + (str == "[" ? "" : ",") + std::to_string(itm);
                }
                return str + "]";
            };
            std::cout << "src shape: " << vec2str(src->getStaticDims()) << ", data ptr = " << src->getData()
                      << std::endl;
            std::cout << "assignedMem shape: " << vec2str(assignedMem->getStaticDims())
                      << ", data ptr = " << assignedMem->getData() << std::endl;
```